### ISSUE TITLE
How to set up a local dev environment

### ISSUE THREAD

**Issue by kytrinyx** [exercism/exercism.io#2095]
See Issue #2092 for an overview of operation welcome contributors.

-----------------------

At the moment the README contains a list of steps to perform, but very little context outside of that. Also, the README probably only works for Mac, can perhaps be used with Linux, and almost certainly does not work on Windows.

Additionally, each language track should have a "how to contribute" section in the README.

From the [OpenHatch guide](http://opensource-events.com/#project-setup):

> Here are common elements of setting up a development environment you’ll want your guide to address:

> **Preparing their computer**
> Make sure they’re familiar with their operating system’s tools, such as the terminal/command prompt. You can do this by linking to a tutorial and asking contributors to make sure they understand it. There are usually great tutorials already out there - OpenHatch’s command line tutorial can be found here.
> If contributors need to set up a virtual environment, access a virtual machine, or download a specific development kit, give them instructions on how to do so.
> List any dependencies needed to run your project, and how to install them. If there are good installation guides for those dependencies, link to them.
> 
> **Downloading the source**
> Give detailed instructions on how to download the source of the project, including common missteps or obstacles.
>
> **How to view/test changes**
> Give instructions on how to view and test the changes they’ve made. This may vary depending on what they’ve changed, but do your best to cover common changes. This can be as simple as viewing an html document in a browser, but may be more complicated.
>
> Installation will often differ depending on the operating system of the contributor. You will probably need to create separate instructions in various parts of your guide for Windows, Mac and Linux users. If you only want to support development on a single operating system, make sure that is clear to users, ideally in the top-level documentation.

**Comment by KaoruDev** [comment]
> very little context outside of that

What else should be mentioned?

**Comment by kytrinyx** [comment]
I'm not sure (that's part of the problem). I suspect that it would be useful to have a high-level overview of what exercism is, what we're trying to achieve, and also how the project is structured (both the different repositories that make up the entire ecosystem, as well as how the main exercism.io code-base is structured), and perhaps also where we intend to go.

I had a brief conversation with a fellow named Bryan Hughes the other day, who is a developer on Johnny Five: https://github.com/rwaldron/johnny-five

He mentioned two things. "gitter" for chat, which is similar to slackchat but with a very tight integration to github, and also:

> We've found that breaking down CONTRIBUTING.md by contributor role rather than technology area (e.g. frontend, cli, database, etc) helps too. Ours is broken down into sections for reporting issues, requesting features, submitting prs, writing tests, writing docs, and creating examples.

**Comment by KaoruDev** [comment]
icic I'll keep this in mind while getting more familiar with the workflow / app

**Comment by kytrinyx** [comment]
That would be great. In particular, keep in mind what questions come up, where you're confused etc, because if you have questions, then I'm pretty sure I could be documenting answers.

**Comment by mtakac** [comment]
First of all, I just recently started contributing to exercism (just the main app for now, but looking forward to move to the API, CLI and some of the lang. tracks as well) and I can say that I haven't got any problems with setup (Linux).

Nevertheless there are few places where I can imagine some improvements. Let me just list some ideas and we can discuss those that will seem interesting in more detail:

1. use some container for development and production. I vouch for Vagrant, but Docker or any similar alternative will do. This solves the problem with folks on Windows and it will very likely save us some headaches in the future - it's really great if all the devs have exactly the same setup as is running on the server.

2. Create a main page for all parts of exercism (I am thinking maybe Github page) with all the vital info for contributors. Also Roadmap would be great - it's really hard to see now where is the project heading and to which parts of app would it be most helpful to contribute. 
Breaking down CONTRIBUTING.md by roles is great but exercism still needs contribution guides by technologyl - somebody working on the frontend really doesn't need to know how the CLI works.

3. Some global overview (could be the part of the project's page) would be really nice. For example some simple diagram showing different pieces and how they come together.

4. Documentation for the API, ideally interactive. Could be maybe https://apiary.io/ or even better, generated from the code. I like http://swagger.io/ and https://github.com/swagger-api/swagger-ui.

5. :+1:  for gitter - it's really cool tool. I like the idea of rooms (One global for whole project and one for each component)

It's probably a completely different issue, but most confusing part for me is the front end. I have no idea what's the plan here and therefore is really hard to help. Apart from that it is quite messy right now, it is also using tools that it doesn't need (E.g. lineman is more-less useless for us right now, if we are using angular, we don't need jquery, mixing coffeescript and javascript also isn't optimal, etc.). The main question here is: Do we want to move in the direction of SPA or do we want templates to be rendered by server and use JS only for some enrichment?

**Comment by kytrinyx** [comment]
This is great feedback, thank you!

I just started looking at docker for development; that seems like it could work beautifully, and is pretty straight-forward to set up.

> most confusing part for me is the front end

Oh, man. Don't get me started. Right now the whole front-end is a mess. Well, more than just the front-end. UX, documentation, GitHub issues. I'm in the middle of doing an audit/cleanup of the issues and creating a roadmap, so I think that we'll be able to tackle some of the bigger issues going forward.

My plan is to move towards more of an SPA approach and separate out the front-end so it lives in it's own repo. There's no need to make front-end contributors deal with Ruby.

Also, as we build out new functionality, I expect to have other stand-alone apps and services (right now rikki- the nitpicking robot is stand-alone, and the problem API.

The next thing is going to be the activity stream stuff which I'll initially build as a Sinatra app, since I have a couple of people who specifically want to help out with it, and Ruby is what they know. If performance becomes a problem we can rebuild just that portion in Go without any issue.

It would be great to work out API documentation for the problem API, but I don't want to document the stuff that lives in the exercism.io repo since every single endpoint is disgustingly ad-hoc at the moment.

The CONTRIBUTING type documentation is what's really had me stuck for a while. I don't know how to organize it, and I really don't know what to put in it. I have so much context living in my head, and I'm not sure what people want/need to know in order to contribute.

**Comment by mtakac** [comment]
Docker is great - easy to use and much faster than Vagrant. Just be aware that it won't solve the problem with development under windows.

I think SPA will be the right move. I have some time next few weeks and I'll be happy to help with cleaning up the front end code, creating proper structure and writing some docs. If you agree I'll open new issue just for discussing the front end.

Activity stream sounds cool and again: Happy to help!

I feel your pain! Coming up with this kind of copy is always the worst part. I'll think about it for a while and let you know any ideas. Maybe we could try to find some other open-source project which handles it well.

**Comment by kytrinyx** [comment]
> Just be aware that it won't solve the problem with development under windows.

Oh, right. Yeah, I forgot about that. On windows you'd need to spin up a VM to run the container, which feels mildly inception-like.

> I'll be happy to help with cleaning up the front end code

That would be amazing! Open up an issue and we can start talking about it.

For the activity stream, I'm breaking down the backend tasks, and was just wondering how the heck I'd get some front-end on it. I'll pull you into that discussion sometime in the next couple of weeks.

For the copy, docker might be a good project to emulate, actually.

**Comment by KaoruDev** [comment]
@mtakac would love it if you could cc me on the front-end issue. I've been meaning to create a POC for a replacement of lineman. I put together a [build tool](https://github.com/kaorudev/jelly) based on gulp. I want to change out some components out of it ( browserify for JSPM or WebPack ) and add a JS test suite plus maybe a end to end suite? Would love to bounce ideas around!

**Comment by mtakac** [comment]
Ok, so I opened new issue dealing with front end: #2296.

@kytrinyx 
> For the activity stream, I'm breaking down the backend tasks, and was just wondering how the heck I'd > get some front-end on it. I'll pull you into that discussion sometime in the next couple of weeks.
Exactly the same way as we'll hook up front end with back end: Activity stream will expose a public API and will serve JSON to the client. Much more complicated (and interesting) will be if you want the stream to be live (which I guess you do, otherwise it wouldn't be call stream, right? :)). Anyway we can discuss this later.

@KaoruDev Sure! Check out the new issue: #2296 and let's discuss things there. The template looks good and I definitely vouch for WebPack

**Comment by kytrinyx** [comment]
> if you want the stream to be live

Ideally yes, but I'm ok with starting with a "feed" variant.

> Activity stream will expose a public API and will serve JSON to the client. 

Yeah, we're on the same page. I'm very relieved that you're willing to help out to get the front-end part sorted out.

**Comment by mtakac** [comment]
No problem! Exercism is really cool and I am happy to help.

**Comment by kickinbahk** [comment]
I went ahead and got exercism.io set up on my machine. I did run into a couple issues here and there and thought that it would be worth discussing how we might want to handle it for newcomers as they are issues that they may often run into. 

The issue that I ran into was related to updating node with homebrew. When I had originally installed node, I did not link it. This made it extremely difficult to update. The fastest way for me to do this was to go through the steps below: 

```sudo brew uninstall node
brew update
brew upgrade
brew cleanup
brew install node
sudo chown -R YourUsernameHere /usr/local
brew link --overwrite node
sudo brew postinstall node```

Lastly, it is imperative that a computer restart is initiated. This is because in postgres update they changed a couple things related to date and time. This was causing it to blow up in all the attempts to migrate the database. 

With all of node's recent changes, I think that it is highly likely for other users to come across these as well. Not sure that we would want to place this in the readme...but maybe link to something about it? I dunno, figured that we can discuss.
